panic while initializing kube-scheduler  if score plugin is disabled and preScore plugins are not.

### What happened?

When initializing kube-scheduler, We provided a custom config file for KubeSchedulerConfiguration and disabled the score plugin.

This resulted in a panic in kube-scheduler at https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/runtime/framework.go#L1034

```
I1027 08:56:23.571036       1 serving.go:376] Generated self-signed cert in-memory
I1027 08:56:23.968849       1 server.go:154] "Starting Kubernetes Scheduler" version="v1.28.3"
I1027 08:56:23.968940       1 server.go:156] "Golang settings" GOGC="" GOMAXPROCS="" GOTRACEBACK=""
I1027 08:56:24.002069       1 secure_serving.go:229] Serving securely on 127.0.0.1:10259
I1027 08:56:24.002670       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I1027 08:56:24.002952       1 shared_informer.go:311] Waiting for caches to sync for RequestHeaderAuthRequestController
I1027 08:56:24.003405       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I1027 08:56:24.003557       1 shared_informer.go:311] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1027 08:56:24.003663       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I1027 08:56:24.003688       1 shared_informer.go:311] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1027 08:56:24.004739       1 tlsconfig.go:345] "Starting DynamicServingCertificateController"
I1027 08:56:24.144610       1 shared_informer.go:318] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1027 08:56:24.204482       1 shared_informer.go:318] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1027 08:56:24.205492       1 leaderelection.go:250] attempting to acquire leader lease kube-system/kube-scheduler...
I1027 08:56:24.205533       1 shared_informer.go:318] Caches are synced for RequestHeaderAuthRequestController
I1027 08:56:39.997587       1 leaderelection.go:260] successfully acquired lease kube-system/kube-scheduler
E1027 06:03:17.384644       1 runtime.go:79] Observed a panic: "makeslice: cap out of range" (runtime error: makeslice: cap out of range)
goroutine 429 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x21ca780?, 0x2878570})
 vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x0?})
 vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x75
panic({0x21ca780, 0x2878570})
 /usr/local/go/src/runtime/panic.go:884 +0x213
k8s.io/kubernetes/pkg/scheduler/framework/runtime.(*frameworkImpl).RunScorePlugins(0xc000343a00, {0x289e3f0, 0xc0003ce2d0}, 0xc000b54fc0, 0xc000a86d80, {0xc000dc0660, 0x3, 0x3})
 pkg/scheduler/framework/runtime/framework.go:1019 +0x173
k8s.io/kubernetes/pkg/scheduler.prioritizeNodes({0x289e3f0, 0xc0003ce2d0}, {0xc0006b8a10, 0x1, 0x1}, {0x28c2908, 0xc000343a00}, 0x1f6795c?, 0xc000a86d80, {0xc000dc0660, ...})
 pkg/scheduler/schedule_one.go:707 +0x1c9
k8s.io/kubernetes/pkg/scheduler.(*Scheduler).schedulePod(0xc00019f800, {0x289e3f0, 0xc0003ce2d0}, {0x28c2908, 0xc000343a00}, 0x58ffd5?, 0xc000a86d80)
 pkg/scheduler/schedule_one.go:406 +0x53d
k8s.io/kubernetes/pkg/scheduler.(*Scheduler).schedulingCycle(0xc00019f800, {0x289e3f0, 0xc0003ce2d0}, 0x2?, {0x28c2908, 0xc000343a00}, 0xc000850000, {0xc146f289543d86d4, 0x44c8dcbef, 0x3ba4de0}, ...)
 pkg/scheduler/schedule_one.go:145 +0xfc
k8s.io/kubernetes/pkg/scheduler.(*Scheduler).scheduleOne(0xc00019f800, {0x289e3f0?, 0xc00039dc70})
 pkg/scheduler/schedule_one.go:107 +0x445
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000066500?, {0x287dd00, 0xc000b57a70}, 0x1, 0xc000a7fb00)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x0, 0x0, 0x40?, 0xc0006a30c8?)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x289e3f0, 0xc00039dc70}, 0xc00040d230, 0x0?, 0x100000000000000?, 0x60?)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x289e3f0?, 0xc00039dc70?}, 0x0?, 0x2662310?)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:170 +0x2b
created by k8s.io/kubernetes/pkg/scheduler.(*Scheduler).Run
 pkg/scheduler/scheduler.go:406 +0xfc
panic: runtime error: makeslice: cap out of range [recovered]
 panic: runtime error: makeslice: cap out of range

goroutine 429 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x0?})
 vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xd7
panic({0x21ca780, 0x2878570})
 /usr/local/go/src/runtime/panic.go:884 +0x213
k8s.io/kubernetes/pkg/scheduler/framework/runtime.(*frameworkImpl).RunScorePlugins(0xc000343a00, {0x289e3f0, 0xc0003ce2d0}, 0xc000b54fc0, 0xc000a86d80, {0xc000dc0660, 0x3, 0x3})
 pkg/scheduler/framework/runtime/framework.go:1019 +0x173
k8s.io/kubernetes/pkg/scheduler.prioritizeNodes({0x289e3f0, 0xc0003ce2d0}, {0xc0006b8a10, 0x1, 0x1}, {0x28c2908, 0xc000343a00}, 0x1f6795c?, 0xc000a86d80, {0xc000dc0660, ...})
 pkg/scheduler/schedule_one.go:707 +0x1c9
k8s.io/kubernetes/pkg/scheduler.(*Scheduler).schedulePod(0xc00019f800, {0x289e3f0, 0xc0003ce2d0}, {0x28c2908, 0xc000343a00}, 0x58ffd5?, 0xc000a86d80)
 pkg/scheduler/schedule_one.go:406 +0x53d
k8s.io/kubernetes/pkg/scheduler.(*Scheduler).schedulingCycle(0xc00019f800, {0x289e3f0, 0xc0003ce2d0}, 0x2?, {0x28c2908, 0xc000343a00}, 0xc000850000, {0xc146f289543d86d4, 0x44c8dcbef, 0x3ba4de0}, ...)
 pkg/scheduler/schedule_one.go:145 +0xfc
k8s.io/kubernetes/pkg/scheduler.(*Scheduler).scheduleOne(0xc00019f800, {0x289e3f0?, 0xc00039dc70})
 pkg/scheduler/schedule_one.go:107 +0x445
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000066500?, {0x287dd00, 0xc000b57a70}, 0x1, 0xc000a7fb00)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x0, 0x0, 0x40?, 0xc0006a30c8?)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x289e3f0, 0xc00039dc70}, 0xc00040d230, 0x0?, 0x100000000000000?, 0x60?)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:259 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x289e3f0?, 0xc00039dc70?}, 0x0?, 0x2662310?)
 vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:170 +0x2b
created by k8s.io/kubernetes/pkg/scheduler.(*Scheduler).Run
 pkg/scheduler/scheduler.go:406 +0xfc
```

### What did you expect to happen?

kube-scheduler up and running

### How can we reproduce it (as minimally and precisely as possible)?

Using the below KubeSchedulerConfiguration file to enable kube-scheduler.

apiVersion: kubescheduler.config.k8s.io/v1
kind: KubeSchedulerConfiguration
clientConnection:
  kubeconfig: /etc/kubernetes/scheduler.conf
profiles:
  - plugins:
      score:
        disabled:
        - name: '*'
extenders:
  - urlPrefix: https://127.0.0.1:12345/scheduler
    filterVerb: "predicates"
    weight: 1
    httpTimeout: 60s
    enableHTTPS: true
    nodeCacheCapable: false
    ignorable: false

### Anything else we need to know?

While trying to debug the issue, I found a change done recently for preScore calculation  which might be causing the issue https://github.com/kubernetes/kubernetes/commit/c0611b6bb331854ba8f2a532248a32f5acce5ff1

When we disable only the Score plugin then scheduler configuration will result in default preScorePlugins with 3 enabled [NodeResourcesFit, InterPodAffinity, TaintToleration]
And due to this change, the preScore for **InterPodAffinity** results in skip status  for some case (still not able to figured out what exactly it is), and  this might result in a panic for kube-scheduler.

```
-------- Line Number 982 framework.go [RunPreScorePlugins]----------
for _, pl := range f.preScorePlugins {
		logger := klog.LoggerWithName(logger, pl.Name())
		ctx := klog.NewContext(ctx, logger)
		status = f.runPreScorePlugin(ctx, pl, state, pod, nodes)
		if status.IsSkip() {
			skipPlugins.Insert(pl.Name())
			continue
		}
		if !status.IsSuccess() {
			return framework.AsStatus(fmt.Errorf("running PreScore plugin %q: %w", pl.Name(), status.AsError()))
		}
	}
state.SkipScorePlugins = skipPlugins
```

The numplugin variable turns out to be negative.  

```
// RunScorePlugins runs the set of configured scoring plugins.
// It returns a list that stores scores from each plugin and total score for each Node.
// It also returns *Status, which is set to non-success if any of the plugins returns
// a non-success status.
func (f *frameworkImpl) RunScorePlugins(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) (ns []framework.NodePluginScores, status *framework.Status) {
startTime := time.Now()
defer func() {
metrics.FrameworkExtensionPointDuration.WithLabelValues(metrics.Score, status.Code().String(), f.profileName).Observe(metrics.SinceInSeconds(startTime))
}()
allNodePluginScores := make([]framework.NodePluginScores, len(nodes))
numPlugins := len(f.scorePlugins) - state.SkipScorePlugins.Len() <====================== L1018
plugins := make([]framework.ScorePlugin, 0, numPlugins). <====================== L1019
pluginToNodeScores := make(map[string]framework.NodeScoreList, numPlugins)
for _, pl := range f.scorePlugins {
if state.SkipScorePlugins.Has(pl.Name()) {
continue
}
plugins = append(plugins, pl)
pluginToNodeScores[pl.Name()] = make(framework.NodeScoreList, len(nodes))
}
```



### Kubernetes version

1.28.3

### Cloud provider

N/A

### OS version

_No response_

### Install tools

kubeadm

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

--- a/pkg/scheduler/framework/runtime/framework.go
+++ b/pkg/scheduler/framework/runtime/framework.go
@@ -1015,7 +1015,7 @@ func (f *frameworkImpl) RunScorePlugins(ctx context.Context, state *framework.Cy
 		metrics.FrameworkExtensionPointDuration.WithLabelValues(metrics.Score, status.Code().String(), f.profileName).Observe(metrics.SinceInSeconds(startTime))
 	}()
 	allNodePluginScores := make([]framework.NodePluginScores, len(nodes))
-	numPlugins := len(f.scorePlugins) - state.SkipScorePlugins.Len()
+	numPlugins := len(f.scorePlugins)
 	plugins := make([]framework.ScorePlugin, 0, numPlugins)
 	pluginToNodeScores := make(map[string]framework.NodeScoreList, numPlugins)
 	for _, pl := range f.scorePlugins {

--- a/pkg/scheduler/framework/runtime/framework_test.go
+++ b/pkg/scheduler/framework/runtime/framework_test.go
@@ -1385,6 +1385,22 @@ func TestRunScorePlugins(t *testing.T) {
 				},
 			},
 		},
+		{
+			name:           "skipped prescore plugin number greater than the number of score plugins",
+			plugins:        buildScoreConfigDefaultWeights(scorePlugin1),
+			pluginConfigs:  nil,
+			skippedPlugins: sets.New(scorePlugin1, "score-plugin-unknown"),
+			want: []framework.NodePluginScores{
+				{
+					Name:   "node1",
+					Scores: []framework.PluginScore{},
+				},
+				{
+					Name:   "node2",
+					Scores: []framework.PluginScore{},
+				},
+			},
+		},
 	}
 
 	for _, tt := range tests {

